hwy/tests/demote_test.cc

```diff
@@ -54,42 +54,106 @@ struct TestDemoteTo {
     auto expected = AllocateAligned<ToT>(N);
 
     // Narrower range in the wider type, for clamping before we cast
-    const T min = LimitsMin<ToT>();
+    const T min = static_cast<T>(IsSigned<T>() ? LimitsMin<ToT>() : ToT{0});
     const T max = LimitsMax<ToT>();
 
-    const auto value_ok = [&](T& value) {
-      if (!IsFiniteT(value)) return false;
-      return true;
-    };
-
     RandomState rng;
     for (size_t rep = 0; rep < AdjustedReps(1000); ++rep) {
       for (size_t i = 0; i < N; ++i) {
-        do {
-          const uint64_t bits = rng();
-          CopyBytes<sizeof(T)>(&bits, &from[i]);  // not same size
-        } while (!value_ok(from[i]));
+        const uint64_t bits = rng();
+        CopyBytes<sizeof(T)>(&bits, &from[i]);  // not same size
         expected[i] = static_cast<ToT>(HWY_MIN(HWY_MAX(min, from[i]), max));
       }
 
       const auto in = Load(from_d, from.get());
       HWY_ASSERT_VEC_EQ(to_d, expected.get(), DemoteTo(to_d, in));
     }
+
+    for (size_t rep = 0; rep < AdjustedReps(1000); ++rep) {
+      for (size_t i = 0; i < N; ++i) {
+        const uint64_t bits = rng();
+        CopyBytes<sizeof(ToT)>(&bits, &expected[i]);  // not same size
+
+        if(!IsSigned<T>() && IsSigned<ToT>())
+          expected[i] &= static_cast<ToT>(max);
+
+        from[i] = static_cast<T>(expected[i]);
+      }
+
+      const auto in = Load(from_d, from.get());
+      HWY_ASSERT_VEC_EQ(to_d, expected.get(), DemoteTo(to_d, in));
+    }
   }
 };
 
 HWY_NOINLINE void TestAllDemoteToInt() {
-  ForDemoteVectors<TestDemoteTo<uint8_t>>()(int16_t());
-  ForDemoteVectors<TestDemoteTo<uint8_t>, 2>()(int32_t());
+  const ForDemoteVectors<TestDemoteTo<uint8_t>> from_i16_to_u8;
+  from_i16_to_u8(int16_t());
+  from_i16_to_u8(uint16_t());
+
+  const ForDemoteVectors<TestDemoteTo<int8_t>> from_i16_to_i8;
+  from_i16_to_i8(int16_t());
+  from_i16_to_i8(uint16_t());
+
+  const ForDemoteVectors<TestDemoteTo<uint8_t>, 2> from_i32_to_u8;
+  from_i32_to_u8(int32_t());
+  from_i32_to_u8(uint32_t());
+
+  const ForDemoteVectors<TestDemoteTo<int8_t>, 2> from_i32_to_i8;
+  from_i32_to_i8(int32_t());
+  from_i32_to_i8(uint32_t());
+
+#if HWY_HAVE_INTEGER64
+#if HWY_HAVE_SCALABLE
```

Comment:
I'm curious why we want to check HWY_HAVE_SCALABLE here? It's intended to be legit to set kPow2=3 in both fixed-size and scalable targets.

Reply:
The reason why HWY_HAVE_SCALABLE is here is because the code was failing to compile on the HWY_RVV target when I originally wrote the unit tests.

Reply:
Interesting. I will investigate after this is merged.